ENH: Standalone benchmark script for the inner loops of ufunc

seiko2plus opened the pull request:
#### ENH: A standalone benchmark script for the inner loops of ufunc
    
This script only measuring the performance of inner loops of ufunc,
the idea behind it is to remove umath object calls from the equation,
in order to reduce the number of noises and provides stable ratios.

eric-wieser commented on benchmarks/misc/benchin_ufunc.py:
```diff
@@ -0,0 +1,273 @@
+class Timing:
+    @staticmethod
+    def set_affinity(*CPUs):
+        utests.ctiming_set_affinity((*CPUs,))
```
```suggestion
        utests.ctiming_set_affinity(tuple(CPUs))
```
is clearer

eric-wieser commented on benchmarks/misc/benchin_ufunc.py:
```diff
@@ -0,0 +1,273 @@
+try:
+    from prettytable import PrettyTable
+except (ImportError, NameError):
+    print("'prettytable' module required for running this script", file=sys.stderr)
+    sys.exit(-1)
```
```suggestion
    raise SystemExit("'prettytable' module required for running this script")
```

eric-wieser commented:
Can we reuse our existing benchmark machinery here?

seiko2plus commented:
@eric-wieser, I tried to use ASV but the result wasn't stable enough, check this [patch](https://github.com/numpy/numpy/pull/13516/commits/59c816a7632f3badb70914300ac09620a3c81a0f) and [patch2](https://github.com/numpy/numpy/pull/13516/commits/8184b71a5776a15be317464f9a1b35f1042b0268) from #13516, the idea behind this patch is to benchmarking only the inner loop of ufunc in order to reduce the noises as much as possible

eric-wieser commented:
It would be nice if we could at least hook into ASV for things like benchmark result comparisons and storage, rather than building our own version of those too. It might be worth starting a conversation with @pv about the best way to do that.

seberg commented:
@seiko2plus you are repeating the function run multiple times here within your `run` function. May that be enough to stabilize the results a bit in asv?

For example, if you just define the `run` function in C (and monkeypatch it into Benchmark), and make it do a couple of C-level calls (to offset the ~200ns or so overhead. That might be enough to get a stable result as well?

seiko2plus commented:
@seberg, ASV already collect multiple samples for each benchmark, but still not stable enough even on idle CPU.

This script is not providing a replacement for the current ASV implementation, the main reason behind it is to detect any performance changes in the inner loops of ufunc and removing the functionality of  umath and multiarry from the equation in order to reduce the noises as much as possible, it also provides more testing cases like multiple strides, sizes and better control for the testing process.

> For example, if you just define the run function in C (and monkeypatch it into Benchmark), and make it do a couple of C-level calls (to offset the ~200ns or so overhead. That might be enough to get a stable result as well?

The problem is ASV doesn't provide a way to specify the elapsed time manually.

seiko2plus commented:
> EDIT: This got lost: "You are doing a few other things here that you are not doing in the asv version."

@seberg, I moved the mentioned patches from #13516, into a separate pull #15992 and #15990. also modified the number for repeats and samples to be equal to the default settings of this script.
but still, the ratio of ASV still not stable enough.

r-devulap commented:
One reason that could be causing noise is turbo mode. In case you haven't already done, I would recommend disabling for benchmarking purposes (set `/sys/devices/system/cpu/intel_pstate/no_turbo `to 1). May be that will help? I haven't had too much  variability while benchmarking ufunc's with `asv`.

seiko2plus commented:
@r-devulap, Before I run any benchmarks, I usually do:
 - isolate logical cores from scheduling through linux kernel options `isolcpus` and `rcu_nocbs`
 - reducing scheduling-clock ticks through `nohz_full` for the isolated cores
 - use option `--cpu-affinity` that comes with this script or what ASV provides for the isolated cores
 - use scaling governor performance via ` /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor`
 - disable turbo boost via `/sys/devices/system/cpu/intel_pstate/no_turbo`
 - make sure that `ASLR`(address space layout randomization) state is 'full randomization' through
   `set 2  to /proc/sys/kernel/randomize_va_space`

Lately, I realized a python module called [pyperf](https://github.com/vstinner/pyperf), provides a tool to tune the system with the above tips and many more via command `pyperf system tune`

However, it seems I should have an idle hardware in order to get almost stable ratios for ASV not just isolate some logical cores since any involved system calls that interpret the thread during collecting the benchmark samples will eliminate the benefits from isolating the logical cores via  `isolcpus` and `rcu_nocbs`.

One of the things I don't like in ASV that its uses a separate process for each collected sample,
which makes it too slow.

mattip commented:
ping @pv. Is there something here that we all are missing?

seiko2plus committed:
ENH: Standalone benchmark script for the inner loops of ufunc

    This script only measuring the performance of inner loops
    of ufunc, the idea behind it is to remove umath object calls
    from the equation, in order to reduce the number of noises and
    provides stable ratios.

hameerabbasi commented:
I ran this PR on a live environment without a desktop (Ubuntu Server), using the method in the PR description. The noise was around 3% and this PR had a performance impact of ±5%, so not too much of a difference.

seiko2plus committed:
improve argument parsing and add new option `--rand-range`

seiko2plus committed:
print numpy info

charris commented:
close/reopen
